MovieHacker/TableViewController.swift

```diff
@@ -0,0 +1,96 @@
+//
+//  ViewController.swift
+//  MovieHacker
+//
+//  Created by Jan Dědeček on 07/08/2017.
+//  Copyright © 2017 Jan Dědeček. All rights reserved.
+//
+
+import UIKit
+
+let apiKey = ""
+
+class TableViewController: UITableViewController {
+
+  var task: URLSessionDataTask?
+
+  var model = [Movie]()
+
+  let imageCache = NSCache<NSString, UIImage>()
+
+  let taskCache = NSCache<NSString, URLSessionDataTask>()
+
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    // Do any additional setup after loading the view, typically from a nib.
+    tableView.cellLayoutMarginsFollowReadableWidth = true
+
+    let today = dateFormatter.string(from: Date())
+    let oneMonthLater = dateFormatter.string(from: Date().addingTimeInterval(3600*24*30))
+
+    //let url = URL(string: "https://api.themoviedb.org/3/movie/upcoming?api_key=\(apiKey)&region=us")!
+    let url = URL(string: "https://api.themoviedb.org/3/discover/movie"
+      + "?primary_release_date.gte=\(today)"
+      + "&primary_release_date.lte=\(oneMonthLater)"
+      + "&api_key=\(apiKey)&region=us&with_release_type=3&sort_by=primary_release_date.asc")!
+
+    task = URLSession.shared.dataTask(with: url) { [weak self] (data, response, error) in
+      if let data = data,
+        let json = try? JSONSerialization.jsonObject(with: data),
+        let model = parse(json: json) {
+        DispatchQueue.main.async {
+          self?.model = model
+          self?.tableView.reloadData()
+        }
+      }
+    }
+    task?.resume()
+  }
+
+  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
```

Comment:
again, this is academic code, so i'd strongly favor a singleton over your ABB pattern, or at least explain in a comment that "We're using dependency injection, because we don't like singletons". Or something...

Reply:
I would I like to do it as we usually do it. Singleton is not really easy in this example.